samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesStoreImpl.java

```diff
@@ -0,0 +1,200 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.samza.operators.impl.store;
+
+import org.apache.samza.storage.kv.ClosableIterator;
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.storage.kv.KeyValueIterator;
+import org.apache.samza.storage.kv.KeyValueStore;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Provides a view on top of a {@link KeyValueStore} that allows retrieval of entries by time ranges.
+ *
+ * <p>
+ * A {@link TimeSeriesStoreImpl} can be backed by persistent stores like rocksDB, in-memory stores, change-logged
+ * stores, cached stores (or any combination of these).
+ *
+ * <p>
+ * Range iterators in the store return values in the order of their timestamp. Within the same key and timestamp,
+ * values are returned in their order of insertion.
+ *
+ * <p>
+ * This store has two modes of operation depending on how duplicates are handled:
+ * <ol>
+ *   <li>
+ *     Overwrite Mode: In this mode, the store only retains the most recent value for a given key and timestamp. I.e.,Calling
+ *     {@link #put} on an existing key and timestamp will overwrite the previously stored value for that key.
+ *   </li>
+ *   <li>
+ *     Append Mode: In this mode, the store retains all previous values for a given key and timestamp. I.e., Calling {@link #put}
+ *     with an existing key and timestamp will append the value to the list.
+ *   </li>
+ * </ol>
+ * <p>
+ * Implementation Notes:
+ *
+ * Data is serialized and organized into K-V pairs as follows:
+ *  <pre>
+ *    +-----------------------+------------------+------------+------------------------+-----------------------+
+ *    |  serialized-key bytes |  timestamp       | version    |       sequence number  | serialized value      |
+ *    |                       |                  |            |                        |                       |
+ *    +-----------------------+------------------+------------+------------------------+-----------------------+
+ *    +----------------------+--------8 bytes----+----1 bytes-+---------7 bytes--------+----value size----------
+ *    +----------------------------------STORE KEY-------------------------------------+---STORE VAL-----------+
+ *  </pre>
+ *  An 8 byte timestamp, a one byte version and a 7 byte sequence number are appended to the provided key and this
+ *  combination is used as the key in the k-v store. The provided value is stored as is.
+ *
+ *  <p> This class is thread-safe and concurrent reads/writes are expected.
```

Comment:
Does the KVStore API/impl provide read after write consistency across threads as well? I.e. do we have any guarantees  about when a value written by T1 will be visible to T2?

Reply:
The RocksDb docs state that the iterators are "thread-safe". We'll have to find out more on whether they mean: 
- Concurrent operations on the store will not corrupt state
- Operations on the store "happen-before" other operations? That directly translates to visibility across threads.

[2] is a much stronger guarantee on [1]